### PR TITLE
add `ilab model list`

### PR DESCRIPTION
**Pull request by cdoern**
`ilab model list` prints a table of all models in ~/.local/share/instructlab/models. Both GGUF and Safetensor models. This cmd also takes a `--list-checkpoints` flag that will expose the checkpoints generated by `ilab model train`. These live in ~/.cache/instructlab/checkpoints

also adds an e2e test checking functionality of gguf and safetensor models

an example output of this command looks like:

```
+------------------------------+---------------------+----------+
| Model Name                   | Last Modified       | Size     |
+------------------------------+---------------------+----------+
| merlinite-7b-lab-Q4_K_M.gguf | 2024-07-24 16:01:57 | 4.1 GiB  |
| instructlab/granite-7b-lab   | 2024-07-24 16:17:39 | 12.6 GiB |
+------------------------------+---------------------+----------+

```

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: scripts/basic-workflow-tests.sh
Comment (nathan-weinberg) [inline scripts/basic-workflow-tests.sh:168]:
```suggestion
    # irregardless, download merlinte safetensors to test that capability
    ilab model download --repository instructlab/merlinite-7b-lab
    ilab model list | grep instructlab/merlinite-7b-lab
```

File: src/instructlab/model/list.py
Comment (nathan-weinberg) [inline src/instructlab/model/list.py:45]: Think we should use `is_model_gguf` here: https://github.com/instructlab/instructlab/blob/main/src/instructlab/model/backends/backends.py#L159

File: src/instructlab/model/list.py
Comment (nathan-weinberg) [inline src/instructlab/model/list.py:39]: The type hint suggest this should be passed as a string, but the default val appears to be a list?

File: scripts/basic-workflow-tests.sh
Comment (cdoern) [reply scripts/basic-workflow-tests.sh:168]: keeping regardless lol

File: requirements.txt
Comment (RobotSail) [inline requirements.txt:48]: What is the impact of adding this dependency - is this something that we've cleared?

File: src/instructlab/model/list.py
Comment (RobotSail) [inline src/instructlab/model/list.py:55]: It seems like there's a lot happening in this `elif` branch and it's leading to a lot of visual clutter. I recommend moving this out to a secondary function since it contains a nested for loop w/ a try/catch.

File: src/instructlab/model/list.py
Comment (RobotSail) [inline src/instructlab/model/list.py:78]: You don't have to change this but here's a tip I like to share since it leads to cleaner branching logic:
```py
# invert the condition
if not list_checkpoints or dir is not DEFAULTS.CHECKPOINTS_DIR:
    continue

# then you can continue here in the main
logging.debug("Including model regardless of ModelCard validity")
```

**Review by RobotSail — COMMENTED** [review]
A few things should be addressed but otherwise LGTM

File: src/instructlab/model/list.py
Comment (tiran) [inline src/instructlab/model/list.py:37]: Type and default should not be needed for a flag
```suggestion
    is_flag=True,
```

File: src/instructlab/model/list.py
Comment (tiran) [inline src/instructlab/model/list.py:59]: pathlib objects have a `walk` method and `resolve` method. The following change should simplify your code. Then you can replace `parts = normalized_path.split(os.sep)` with `root.parts`.

```suggestion
                for root, _, files in entry.resolve().walk():
```

File: src/instructlab/model/list.py
Comment (tiran) [reply src/instructlab/model/list.py:54]: That's an excellent suggestion. I also recommend to move if/elif block into a separate private function that returns name, age, and size. It may even a good idea to move each analyze code into its own private function, too. That would make it easier to write unit tests for each implementation.

```python
AnalyzeResult = tuple[str, str, str]
def _analyze_gguf(entry: pathlib.Path) -> AnalyzeResult:
    ...
```

File: src/instructlab/model/list.py
Comment (tiran) [inline src/instructlab/model/list.py:51]: ctime is probably the wrong thing. It's the time stamp of the last status change of the file's metadata (owner, permission, ...). mtime is the last file data modification time.

**Review by tiran — DISMISSED** [review]

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:51]: ah cool, thanks!

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:59]: thanks for the tip! will change accordingly

Comment (cdoern) [reply src/instructlab/model/list.py:59]: getting `AttributeError: 'PosixPath' object has no attribute 'walk'`

Comment (tiran) [reply src/instructlab/model/list.py:59]: argh, my bad. The method was added in 3.12. We have to support 3.10 and 3.11.

File: src/instructlab/model/list.py
Comment (jaideepr97) [inline src/instructlab/model/list.py:118]: what if we just reused the `is_model_safetensors()` check to only list out valid models? 
If we must use the model card method maybe that should be added to the `is_model_safetensors()` check so we have one comprehensive check for how we decide something is a valid model

**Review by jaideepr97 — COMMENTED** [review]
Thanks for the PR @cdoern 
Please update the changelog as well since this is new functionality

File: src/instructlab/model/list.py
Comment (RobotSail) [reply src/instructlab/model/list.py:118]: Wouldn't we also want to include llama-cpp models for the time being as well?

Comment (jaideepr97) [reply src/instructlab/model/list.py:118]: @RobotSail there's a separate check for gguf models, this block only checks against model paths that are directories

Comment (cdoern) [reply src/instructlab/model/list.py]: switched it

File: src/instructlab/model/list.py
Comment (RobotSail) [inline src/instructlab/model/list.py:67]: The way you're printing the formatted headers is very smart. You can also optimize this loop even further by using `zip` to go through both iterators:


```suggestion
    for header, width in zip(headers, column_widths):
        print(f"{item:width}", end=" |)
    print()
```

File: src/instructlab/model/list.py
Comment (RobotSail) [inline src/instructlab/model/list.py:63]: I'm not sure how to suggest this big of a change and don't feel like making a separate PR for this, but I've fixed up how we're printing this so now it's printing in the format we originally wanted from PrettyTable:

```
+--------------------------------+---------------------------+-----------------------+
| Model Name                     | Model Age                 | Model Size            |
+--------------------------------+---------------------------+-----------------------+
| merlinite-7b-lab-Q4_K_M-2.gguf | 0 days 00:35:09.434472546 | 4.068468689918518 GB  |
| instructlab/granite-7b-lab     | 0 days 00:32:32.872044662 | 12.553834554739296 GB |
| merlinite-7b-lab-Q4_K_M.gguf   | 0 days 00:40:45.984725562 | 4.068468689918518 GB  |
+--------------------------------+---------------------------+-----------------------+
```


```suggestion
# print_table displays the models on the system in a nicely formatted table
def print_table(headers: list[str], data: list[list[str]]):
    column_widths = [
        max(len(str(row[i])) for row in data + [headers]) for i in range(len(headers))
    ]

    # Print separator line between headers and data
    horizontal_lines = ["-" * (width + 2) for width in column_widths]
    joining_line = "+" + "+".join(horizontal_lines) + "+"

    print(joining_line)

    outputs = []
    for header, width in zip(headers, column_widths, strict=False):
        outputs.append(f" {header:{width}} ")
    print("|" + "|".join(outputs) + "|")

    print(joining_line)

    for row in data:
        outputs = []
        for item, width in zip(row, column_widths, strict=False):
            outputs.append(f" {item:{width}} ")
        print("|" + "|".join(outputs) + "|")

    print(joining_line)```

**Review by RobotSail — APPROVED** [review]
Really awesome work on this PR Charlie, this will be a great add. 

The only thing we should do before merging is fixing the `Model Age` field and `Model Size` field so that:
- Model Size is rounded to 2 decimal points (currently we're printing full precision)
- Model Age should be renamed to "Downloaded At" because that's what it seems to be representing right now
- For the values of model age, we should change it in one of the two following ways:
  - Print out the most significant unit like how podman does, e.g.: `2 seconds ago`, `5 minutes ago`, `1 hour ago`, etc.
  - OR just print out a simple timestamp in a format like mm:dd:yy, hh:mm 

I also provided a snippet in the review to polish up the output a little bit to resemble PrettyTable. 

But I don't want to hold this PR up too much, so we should get whatever the simplest fixes are in and merge this. 

Thank you for putting this together, this is great work!

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:51]:
```suggestion
    magnitudes = ["KiB", "MiB", "GiB", "TiB"]
```

Since you are using 1024 you are not printing Gygabyte but Gigibyte and so on. Let's add TiB too.

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:91]:
```suggestion
    return [entry.name, str(age), f"{f_size:.1f} {magnitude}"]
```

You don't need full precision, printing `4.1 GiB` is more than enough.

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:90]: Why do we need a nanosecond precision here? This seems quite overkill.

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:11]: Panda is not a small library, it took a few seconds for me to load it. Are we ok adding that up to the CLI?

**Review by leseb — CHANGES_REQUESTED** [review]

File: src/instructlab/model/backends/backends.py
Comment (jaideepr97) [reply src/instructlab/model/backends/backends.py:183]: for me as well

Comment (cdoern) [reply src/instructlab/model/backends/backends.py:183]: spacing is what It should be I believe. I have noticed this recently with GH reviews representing changes like this.

File: src/instructlab/model/list.py
Comment (jaideepr97) [inline src/instructlab/model/list.py:115]: please update comment to remove ModelCard reference on 114
Also wondering if we need the `is_model_gguf` check again here? seems redundant

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:90]: switched, also removed pandas

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:11]: removed, sorry!

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:115]: we need gguf check probably, there could be nested dirs.

File: src/instructlab/model/list.py
Comment (cdoern) [reply src/instructlab/model/list.py:62]: thanks, incorporated

Comment (cdoern) [reply src/instructlab/model/list.py:62]: this is REALLY nice

File: src/instructlab/model/backends/backends.py
Comment (leseb) [inline src/instructlab/model/backends/backends.py:175]: Can you make sure this is not a space/tab issue and stick with spaces? Thanks

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:48]:
```suggestion
# convert_bytes_to_proper_mag takes a dir/file size in Bytes and converts it to the proper Magnitude (MiB, GiB)
```

File: src/instructlab/model/backends/backends.py
Comment (cdoern) [reply src/instructlab/model/backends/backends.py:175]: when I edit the file, changing to spaces results in no changes to the file to commit, same goes for tabs, unsure how to fix how this is rendering

**Comment by cdoern** [comment]
just fixed the MiB and GiB comment and pushed it up

**Comment by cdoern** [comment]
*shakes fist at changelog*

File: src/instructlab/model/list.py
Comment (leseb) [inline src/instructlab/model/list.py:100]: do we follow symlinks?

Comment (cdoern) [reply src/instructlab/model/list.py:100]: nope `stat = Path(full_file).stat(follow_symlinks=False)`

I think by default `os.walk` doesn't either?

Comment (leseb) [reply src/instructlab/model/list.py:100]: Aren't we using symlinks from cache to model dir? @RobotSail do you remember? Thanks

Comment (jaideepr97) [reply src/instructlab/model/list.py:100]: OCI downloaded models are also symlinked from cache to model dir 
do we need to follow them to get metadata about the file, or is that stuff carried over in the symlink?

Comment (cdoern) [reply src/instructlab/model/list.py:100]: cc @mbestavros

Comment (cdoern) [reply src/instructlab/model/list.py:100]: speaking to oleg and mark now, they will follow up here

Comment (mbestavros) [reply src/instructlab/model/list.py:100]: @RobotSail and I hacked a bit on model paths last week - IIRC the model loaders ran into issues when we provided symlinks to them. I'd verify that but that would be the strongest reason I think

Comment (cdoern) [reply src/instructlab/model/list.py:100]: I think we should get this in as is, and follow up with enhancements IMO

**Comment by booxter** [comment]
@cdoern I didn't realize that this work is also included in another PR and left a bunch of comments that apply here [elsewhere](https://github.com/instructlab/instructlab/pull/1861#pullrequestreview-2199749611). Would you mind checking the comments there so that I don't have to copy them verbatim? Thanks!

**Comment by cdoern** [comment]
> @cdoern I didn't realize that this work is also included in another PR and left a bunch of comments that apply here [elsewhere](https://github.com/instructlab/instructlab/pull/1861#pullrequestreview-2199749611). Would you mind checking the comments there so that I don't have to copy them verbatim? Thanks!

The review on the other PR is very eleventh hour in the sense that this PR has two approvals and is about to merge. I think if anything it would make sense to follow up in a different PR and/or include these changes in @RobotSail's `ilab data list` PR, though it is a separate cmd. 

I do really appreciate the thoroughness of the review, and if you want can make an immediate follow up. We would really like to get this functionality into the beta release today

**Comment by booxter** [comment]
@cdoern I can send a follow-up, sure.

File: scripts/basic-workflow-tests.sh
Comment (nathan-weinberg) [inline scripts/basic-workflow-tests.sh:158]:
```suggestion
        step Listing the mixtral GGUF model only
```

File: scripts/basic-workflow-tests.sh
Comment (nathan-weinberg) [inline scripts/basic-workflow-tests.sh:165]: from my last review
```suggestion
    # regardless, download merlinite safetensors to test that capability
```

File: src/instructlab/model/backends/backends.py
Comment (nathan-weinberg) [inline src/instructlab/model/backends/backends.py:186]:
```suggestion
        logger.debug(f"'model_path' is a directory, returning {exc}")
```

**Review by nathan-weinberg — APPROVED** [review]
Couple nits but otherwise LGTM

**Comment by mairin** [comment]
@tiran Just a note I gave Charlie the ok to remove request for review; the requests were addressed as asked and a while had passed and I felt we had to get into today's beta because it was blocking another PR. My apologies if I gave bad guidance on this.

**Pull request merged**
